README.md
# Larynx

🎥 [DEMO VIDEO](https://www.youtube.com/watch?v=hBmhDf8cl0k)

Offline end-to-end text to speech system using [gruut](https://github.com/rhasspy/gruut) and [onnx](https://onnx.ai/) ([architecture](#architecture)). There are [50 voices available across 9 languages](#samples).

```sh
curl https://raw.githubusercontent.com/rhasspy/larynx/master/docker/larynx-server \
    > ~/bin/larynx-server && chmod +755 ~/bin/larynx-server
larynx-server
```

Visit http://localhost:5002 for the test page. See http://localhost:5002/openapi/ for HTTP endpoint documentation.

![Larynx screenshot](img/web_screenshot.png)

Supports a [subset of SSML](#ssml) that can use multiple voices and languages!

``` xml
<speak>
  The 1st thing to remember is that 9 languages are supported in Larynx TTS as of 10/19/2021 at 10:39am.

  <voice name="harvard">
    <s>
      The current voice can be changed!
    </s>
  </voice>

  <voice name="northern_english_male">
    <s>Breaks are possible</s>
    <break time="0.5s" />
    <s>between sentences.</s>
  </voice>

  <s lang="en">
    One language is never enough
  </s>
  <s lang="de">
   Eine Sprache ist niemals genug
  </s>
  <s lang="sw">
    Lugha moja haitoshi
  </s>
</speak>
```

Larynx's goals are:

* "Good enough" synthesis to avoid using a cloud service
* Faster than realtime performance on a Raspberry Pi 4 (with low quality vocoder)
* Broad language support (9 languages)
* Voices trained purely from public datasets

You can use Larynx to:

* Host a [text to speech HTTP endpoint](#docker-installation)
* Synthesize text [on the command-line](#basic-synthesis)
* [Read a book](#long-texts) to you

## Samples

[Listen to voice samples](https://rhasspy.github.io/larynx/) from all of the [pre-trained voices](https://github.com/rhasspy/larynx/releases).

---

## Docker Installation

Pre-built Docker images are available for the following platforms:

* `linux/amd64` - desktop/laptop/server
* `linux/arm64` - Raspberry Pi 64-bit
* `linux/arm/v7` - Raspberry Pi 32-bit

These images include a single English voice, but [many more can be downloaded](https://github.com/rhasspy/larynx/releases/tag/2021-03-28) from within the web interface.

The [larynx](https://raw.githubusercontent.com/rhasspy/larynx/master/docker/larynx) and [larynx-server](https://raw.githubusercontent.com/rhasspy/larynx/master/docker/larynx-server) shell scripts wrap the Docker images, allowing you to use Larynx as a command-line tool.

To manually run the Larynx web server in Docker:

```sh
docker run \
    -it \
    -p 5002:5002 \
    -e "HOME=${HOME}" \
    -v "$HOME:${HOME}" \
    -v /usr/share/ca-certificates:/usr/share/ca-certificates \
    -v /etc/ssl/certs:/etc/ssl/certs \
    -w "${PWD}" \
    --user "$(id -u):$(id -g)" \
    rhasspy/larynx
```

Downloaded voices will be stored in `${HOME}/.local/share/larynx`.

Visit http://localhost:5002 for the test page. See http://localhost:5002/openapi/ for HTTP endpoint documentation.

## Debian Installation

Pre-built Debian packages for [bullseye](https://www.debian.org/releases/bullseye/) are [available for download](https://github.com/rhasspy/larynx/releases/latest) with the name `larynx-tts_<VERSION>_<ARCH>.deb` where `ARCH` is one of `amd64` (most desktops, laptops), `armhf` (32-bit Raspberry Pi), and `arm64` (64-bit Raspberry Pi)
    
Example installation on a typical desktop:

```sh
sudo apt install ./larynx-tts_<VERSION>_amd64.deb
```

From there, you may run the `larynx` command or `larynx-server` to start the web server (http://localhost:5002).

## Python Installation

You may need to install the following dependencies (besides Python 3.7+):

```sh
sudo apt-get install libopenblas-base libgomp1 libatomic1
```

On 32-bit ARM systems (Raspberry Pi), you will also need:

```sh
sudo apt-get install libatlas3-base libgfortran5
```

Next, create a Python virtual environment:

```sh
python3 -m venv larynx_venv
source larynx_venv/bin/activate

pip3 install --upgrade pip
pip3 install --upgrade wheel setuptools
```

Next, install larynx:

```sh
pip3 install -f 'https://synesthesiam.github.io/prebuilt-apps/' -f 'https://download.pytorch.org/whl/cpu/torch_stable.html' larynx
```

Then run `larynx` or `larynx.server` for the web server. You may also execute the Python module directly with `python3 -m larynx` and `python3 -m larynx.server`.

### Voice/Vocoder Download

Voices and vocoders are automatically downloaded when used on the command-line or in the web server. You can also [manually download each voice](https://github.com/rhasspy/larynx/releases/tag/2021-03-28). Extract them to `${HOME}/.local/share/larynx/voices` so that the directory structure follows the pattern `${HOME}/.local/share/larynx/voices/<language>,<voice>`.

---

## Command-Line Interface

Larynx has a flexible command-line interface, available with:

* The [larynx script](https://raw.githubusercontent.com/rhasspy/larynx/master/docker/larynx) for Docker 
* The `larynx` command from the Debian package
* `larynx` or `python3 -m larynx` for Python installations

### Basic Synthesis

```sh
larynx -v <VOICE> "<TEXT>" > output.wav
```

where `<VOICE>` is a language name (`en`, `de`, etc) or a voice name (`ljspeech`, `thorsten`, etc). `<TEXT>` may contain multiple sentences, which will be combined in the final output WAV file. These can also be [split into separate WAV files](#multiple-wav-output).

To adjust the quality of the output, use `-q <QUALITY>` where `<QUALITY>` is "high" (slowest), "medium", or "low" (fastest).

### SSML Synthesis

```sh
larynx --ssml -v <VOICE> "<SSML>" > output.wav
```

where `<SSML>` is valid [SSML](https://www.w3.org/TR/speech-synthesis11/). Not all features are supported; for example:

* Breaks (pauses) can only occur between sentences and can only be specified in seconds or milliseconds
* Voices can only be referenced by name
* Custom lexicons are not yet supported (you can use `<phoneme ph="...">`, however)

If your SSML contains `<mark>` tags, add `--mark-file <FILE>` to the command-line. As the marks are encountered (between sentences), their names will be written on separate lines to the file.

### CUDA Accelerated Synthesis

The `--cuda` flag will make use of a GPU if its available to PyTorch:

``` sh
larynx --cuda 'This is spoken on the GPU.' > output.wav
```

Adding the `--half` flag will enable half-precision inference, which is often faster:

``` sh
larynx --cuda --half 'This is spoken on the GPU even faster.' > output.wav
```

For CUDA acceleration to work, your voice must contain a PyTorch checkpoint file (`generator.pth`). Older Larynx voices did not have these, so you may need to [re-download your voices](https://github.com/rhasspy/larynx/releases/latest/).

### Long Texts

If your text is very long, and you would like to listen to it as its being synthesized, use the `--raw-stream` option:

```sh
larynx -v en --raw-stream < long.txt | aplay -r 22050 -c 1 -f S16_LE
```

Each input line will be synthesized and written the standard out as raw 16-bit 22050Hz mono PCM. By default, 5 sentences will be kept in an output queue, only blocking synthesis when the queue is full. You can adjust this value with `--raw-stream-queue-size`. Additionally, you can adjust `--max-thread-workers` to change how many threads are available for synthesis.

If your long text is fixed-width with blank lines separating paragraphs like those from [Project Gutenberg](https://www.gutenberg.org/), use the `--process-on-blank-line` option so that sentences will not be broken at line boundaries. For example, you can listen to "Alice in Wonderland" like this:

```sh
curl --output - 'https://www.gutenberg.org/files/11/11-0.txt' | \
    larynx -v ek --raw-stream --process-on-blank-line | aplay -r 22050 -c 1 -f S16_LE
```

### Multiple WAV Output

With `--output-dir` set to a directory, Larynx will output a separate WAV file for each sentence:

```sh
larynx -v en 'Test 1. Test 2.' --output-dir /path/to/wavs
```

By default, each WAV file will be named using the (slightly modified) text of the sentence. You can have WAV files named using a timestamp instead with `--output-naming time`. For full control of the output naming, the `--csv` command-line flag indicates that each sentence is of the form `id|text` where `id` will be the name of the WAV file.

```sh
cat << EOF |
s01|The birch canoe slid on the smooth planks.
s02|Glue the sheet to the dark blue background.
s03|It's easy to tell the depth of a well.
s04|These days a chicken leg is a rare dish.
s05|Rice is often served in round bowls.
s06|The juice of lemons makes fine punch.
s07|The box was thrown beside the parked truck.
s08|The hogs were fed chopped corn and garbage.
s09|Four hours of steady work faced us.
s10|Large size in stockings is hard to sell.
EOF
  larynx --csv --voice en --output-dir /path/to/wavs
```

### Interactive Mode

With no text input and no output directory, Larynx will switch into interactive mode. After entering a sentence, it will be played with `--play-command` (default is `play` from SoX).

```sh
larynx -v en
Reading text from stdin...
Hello world!<ENTER>
```

Use `CTRL+D` or `CTRL+C` to exit.

### GlowTTS Settings

The GlowTTS voices support two additional parameters:

* `--noise-scale` - determines the speaker volatility during synthesis (0-1, default is  0.667)
* `--length-scale` - makes the voice speaker slower (> 1) or faster (< 1)

### Vocoder Settings

* `--denoiser-strength` - runs the denoiser if > 0; a small value like 0.005 is a good place to start.

### List Voices and Vocoders

```sh
larynx --list
```

---

## MaryTTS Compatible API

To use Larynx as a drop-in replacement for a [MaryTTS](http://mary.dfki.de/) server (e.g., for use with [Home Assistant](https://www.home-assistant.io/integrations/marytts/)), run:

```sh
docker run \
    -it \
    -p 59125:5002 \
    -e "HOME=${HOME}" \
    -v "$HOME:${HOME}" \
    -v /usr/share/ca-certificates:/usr/share/ca-certificates \
    -v /etc/ssl/certs:/etc/ssl/certs \
    -w "${PWD}" \
    --user "$(id -u):$(id -g)" \
    rhasspy/larynx
```

The `/process` HTTP endpoint should now work for voices formatted as `<LANG>` or `<VOICE>`, e.g. `en` or `harvard`.

You can specify the vocoder quality by adding `;<QUALITY>` to the MaryTTS voice where `QUALITY` is "high", "medium", or "low".

For example: `en;low` will use the lowest quality (but fastest) vocoder. This is usually necessary to get decent performance on a Raspberry Pi.

---

## SSML

A subset of [SSML](https://www.w3.org/TR/speech-synthesis11/) is supported (use `--ssml`):

* `<speak>` - wrap around SSML text
    * `lang` - set language for document
* `<s>` - sentence (disables automatic sentence breaking)
    * `lang` - set language for sentence
* `<w>` / `<token>` - word (disables automatic tokenization)
* `<voice name="...">` - set voice of inner text
    * `voice` - name or language of voice
* `<say-as interpret-as="">` - force interpretation of inner text
    * `interpret-as` one of "spell-out", "date", "number", "time", or "currency"
    * `format` - way to format text depending on `interpret-as`
        * number - one of "cardinal", "ordinal", "digits", "year"
        * date - string with "d" (cardinal day), "o" (ordinal day), "m" (month), or "y" (year)
* `<break time="">` - Pause for given amount of time
    * time - seconds ("123s") or milliseconds ("123ms")
* `<mark name="">` - User-defined mark (written to `--mark-file` or part of `TextToSpeechResult`)
    * name - name of mark
* `<sub alias="">` - substitute `alias` for inner text
* `<phoneme ph="...">` - supply phonemes for inner text
    * `ph` - phonemes for each word of inner text, separated by whitespace
* `<lexicon id="...">` - inline pronunciation lexicon
    * `id` - unique id of lexicon (used in `<lookup ref="...">`)
    * One or more `<lexeme>` child elements with:
        * `<grapheme role="...">WORD</grapheme>` - word text (optional [role][#word-roles])
        * `<phoneme>P H O N E M E S</phoneme>` - word pronunciation (phonemes separated by whitespace)
* `<lookup ref="...">` - use inline pronunciation lexicon for child elements
    * `ref` - id from a `<lexicon id="...">`

### Word Roles

During phonemization, word roles are used to disambiguate pronunciations. Unless manually specified, a word's role is derived from its part of speech tag as `gruut:<TAG>`. For initialisms and `spell-out`, the role `gruut:letter` is used to indicate that e.g., "a" should be spoken as `/eɪ/` instead of `/ə/`.

For `en-us`, the following additional roles are available from the part-of-speech tagger:

* `gruut:CD` - number
* `gruut:DT` - determiner
* `gruut:IN` - preposition or subordinating conjunction 
* `gruut:JJ` - adjective
* `gruut:NN` - noun
* `gruut:PRP` - personal pronoun
* `gruut:RB` - adverb
* `gruut:VB` - verb
* `gruut:VB` - verb (past tense)

### Inline Lexicons

Inline [pronunciation lexicons](https://www.w3.org/TR/2008/REC-pronunciation-lexicon-20081014/) are supported via the `<lexicon>` and `<lookup>` tags. gruut diverges slightly from the [SSML standard](https://www.w3.org/TR/speech-synthesis11/) here by only allowing lexicons to be defined within the SSML document itself. Additionally, the `id` attribute of the `<lexicon>` element can be left off to indicate a "default" inline lexicon that does not require a corresponding `<lookup>` tag.

For example, the following document will yield three different pronunciations for the word "tomato":

``` xml
<?xml version="1.0"?>
<speak version="1.1"
       xmlns="http://www.w3.org/2001/10/synthesis"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="http://www.w3.org/2001/10/synthesis
                 http://www.w3.org/TR/speech-synthesis11/synthesis.xsd"
       xml:lang="en-US">

  <lexicon xml:id="test" alphabet="ipa">
    <lexeme>
      <grapheme>
        tomato
      </grapheme>
      <phoneme>
        <!-- Individual phonemes are separated by whitespace -->
        t ə m ˈɑ t oʊ
      </phoneme>
    </lexeme>
    <lexeme>
      <grapheme role="fake-role">
        tomato
      </grapheme>
      <phoneme>
        <!-- Made up pronunciation for fake word role -->
        t ə m ˈi t oʊ
      </phoneme>
    </lexeme>
  </lexicon>

  <w>tomato</w>
  <lookup ref="test">
    <w>tomato</w>
    <w role="fake-role">tomato</w>
  </lookup>
</speak>
```

The first "tomato" will be looked up in the U.S. English lexicon (`/t ə m ˈeɪ t oʊ/`). Within the `<lookup>` tag's scope, the second and third "tomato" words will be looked up in the inline lexicon. The third "tomato" word has a [role](#word-roles) attached  (selecting a made up pronunciation in this case).

Even further from the SSML standard, gruut allows you to leave off the `<lexicon>` id entirely. With no `id`, a `<lookup>` tag is no longer needed, allowing you to override the pronunciation of any word in the document: 

``` xml
<?xml version="1.0"?>
<speak version="1.1"
       xmlns="http://www.w3.org/2001/10/synthesis"
       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
       xsi:schemaLocation="http://www.w3.org/2001/10/synthesis
                 http://www.w3.org/TR/speech-synthesis11/synthesis.xsd"
       xml:lang="en-US">

  <!-- No id means change all words without a lookup -->
  <lexicon>
    <lexeme>
      <grapheme>
        tomato
      </grapheme>
      <phoneme>
        t ə m ˈɑ t oʊ
      </phoneme>
    </lexeme>
  </lexicon>

  <w>tomato</w>
</speak>
```

This will yield a pronunciation of `/t ə m ˈɑ t oʊ/` for all instances of "tomato" in the document (unless they have a `<lookup>`).


---

## Text to Speech Models

* [GlowTTS](https://github.com/rhasspy/glow-tts-train) (50 voices)
    * English (`en-us`, 27 voices)
        * blizzard\_fls (F, accent, [Blizzard](https://www.cstr.ed.ac.uk/projects/blizzard/2017/usborne_blizzard2017/license.html))
        * blizzard\_lessac (F, [Blizzard](https://www.cstr.ed.ac.uk/projects/blizzard/2011/lessac_blizzard2011/license.html))
        * cmu\_aew (M, [Arctic](licenses/cmuarctic.txt))
        * cmu\_ahw (M, [Arctic](licenses/cmuarctic.txt))
        * cmu\_aup (M, accent, [Arctic](licenses/cmuarctic.txt))
        * cmu\_bdl (M, [Arctic](licenses/cmuarctic.txt))

CHANGELOG
## [1.1.0] - 11 Nov 2021

### Changed

- With --ssml, input from stdin is assumed to be one document instead of lines (override with --stdin-format lines)
- Bump gruut to version 2.1 for inline lexicons (https://github.com/rhasspy/gruut/#inline-lexicons)
- Include /usr/share/ca-certificates in Docker scripts

## [1.0.0] - 20 Oct 2021

### Added

- Support for a subset of SSML (--ssml flag)
- Ability to use CUDA (--cuda) and half-precision inference (--half)


larynx/server.py
#!/usr/bin/env python3
"""Larynx web server"""
import argparse
import asyncio
import contextlib
import functools
import io
import logging
import signal
import time
import typing
from pathlib import Path
from urllib.parse import parse_qs
from uuid import uuid4

import gruut_ipa
import hypercorn
import numpy as np
import pidfile
import quart_cors
from quart import (
    Quart,
    Response,
    jsonify,
    render_template,
    request,
    send_from_directory,
)
from swagger_ui import api_doc

from larynx import __version__, text_to_speech
from larynx.constants import VocoderQuality
from larynx.utils import (
    DEFAULT_VOICE_URL_FORMAT,
    VOCODER_DIR_NAMES,
    VOCODER_QUALITY,
    VOICE_DOWNLOAD_NAMES,
    VOICE_GENDERS,
    download_voice,
    get_voices_dirs,
    load_voices_aliases,
    load_voices_genders,
    valid_voice_dir,
)
from larynx.wavfile import write as wav_write

_DIR = Path(__file__).parent
_WAV_DIR = _DIR / "wav"

_LOGGER = logging.getLogger("larynx")
_LOOP = asyncio.get_event_loop()

# language -> file name
_SAMPLE_SENTENCES = {
    "de-de": "haben_sie_ein_vegetarisches",
    "en-us": "it_took_me_quite_a_long_time_to_develop_a_voice",
    "es-es": "siga_recto",
    "fr-fr": "pourriez-vous_parler",
    "it-it": "parli_un_altra",
    "nl": "kunt_u_wat_langzamer_praten_alstublieft",
    "ru-ru": "Моё_судно_на",
    "sv-se": "den_här_damen",
    "sw": "gari_langu_linaloangama_limejaa_na_mikunga",
}

# -----------------------------------------------------------------------------

parser = argparse.ArgumentParser(prog="larynx")
parser.add_argument(
    "--host", default="0.0.0.0", help="Host of HTTP server (default: 0.0.0.0)"
)
parser.add_argument(
    "--port", type=int, default=5002, help="Port of HTTP server (default: 5002)"
)
parser.add_argument(
    "--voices-dir",
    help="Directory with <LANGUAGE>/<VOICE> structure (overrides LARYNX_VOICES_DIR env variable)",
)
parser.add_argument(
    "--quality",
    choices=["high", "medium", "low"],
    default="high",
    help="Vocoder quality used if not set in API call (default: high)",
)
parser.add_argument(
    "--denoiser-strength",
    type=float,
    default=0.005,
    help="Denoiser strength used if not set in API call (default: 0.005)",
)
parser.add_argument(
    "--noise-scale",
    type=float,
    default=0.667,
    help="Noise scale (voice volatility) used if not set in API call (default: 0.667)",
)
parser.add_argument(
    "--length-scale",
    type=float,
    default=1.0,
    help="Length scale (<1 is faster) used if not set in API call (default: 1.0)",
)
parser.add_argument(
    "--url-format",
    default=DEFAULT_VOICE_URL_FORMAT,
    help="Format string for download URLs (accepts {voice})",
)
parser.add_argument(
    "--pidfile", help="Path to pidfile. Exit if pidfile already exists."
)
parser.add_argument("--logfile", help="Path to logging file (default: stderr)")
parser.add_argument("--cuda", action="store_true", help="Use CUDA if available")
parser.add_argument(
    "--half",
    action="store_true",
    help="Use faster FP16 for inference (requires --cuda)",
)
parser.add_argument(
    "--debug", action="store_true", help="Print DEBUG messages to console"
)
args = parser.parse_args()

# Set up logging
log_args = {}

if args.debug:
    log_args["level"] = logging.DEBUG
else:
    log_args["level"] = logging.INFO

if args.logfile:
    log_args["filename"] = args.logfile

logging.basicConfig(**log_args)  # type: ignore
_LOGGER.debug(args)

if args.cuda:
    import torch

    args.cuda = torch.cuda.is_available()
    if not args.cuda:
        args.half = False
        _LOGGER.warning("CUDA is not available")

voices_dirs = get_voices_dirs(args.voices_dir)
load_voices_aliases()


# -----------------------------------------------------------------------------

app = Quart("larynx")
app.secret_key = str(uuid4())

if args.debug:
    app.config["TEMPLATES_AUTO_RELOAD"] = True

app = quart_cors.cors(app)

# -----------------------------------------------------------------------------


async def text_to_wav(
    text: str,
    voice: str,
    vocoder: str,
    denoiser_strength: typing.Optional[float] = None,
    noise_scale: typing.Optional[float] = None,
    length_scale: typing.Optional[float] = None,
    ssml: bool = False,
) -> bytes:
    """Runs TTS for each line and accumulates all audio into a single WAV."""
    # <lang>/<voice> -> <lang>_<voice>
    voice = voice.replace("/", "_")

    # Settings
    tts_settings = None
    if (noise_scale is not None) or (length_scale is not None):
        tts_settings = {}
        if noise_scale is not None:
            tts_settings["noise_scale"] = noise_scale

        if length_scale is not None:
            tts_settings["length_scale"] = length_scale

    vocoder_settings = None
    if denoiser_strength is not None:
        vocoder_settings = {"denoiser_strength": denoiser_strength}

    # Synthesize each line separately.
    # Accumulate into a single WAV file.
    _LOGGER.info("Synthesizing with %s, %s (%s char(s))...", voice, vocoder, len(text))
    start_time = time.time()

    tts_results = await _LOOP.run_in_executor(
        None,
        functools.partial(
            text_to_speech,
            text=text,
            voice_or_lang=voice,
            vocoder_or_quality=vocoder,
            tts_settings=tts_settings,
            vocoder_settings=vocoder_settings,
            use_cuda=args.cuda,
            half=args.half,
            ssml=ssml,
            url_format=args.url_format,
        ),
    )

    audios = []
    sample_rate = 22050
    for result in tts_results:
        sample_rate = result.sample_rate
        audios.append(result.audio)

    with io.BytesIO() as wav_io:
        wav_write(wav_io, sample_rate, np.concatenate(audios))
        wav_bytes = wav_io.getvalue()

    end_time = time.time()
    _LOGGER.info(
        "Synthesized %s byte(s) in %s second(s)", len(wav_bytes), end_time - start_time
    )

    return wav_bytes


def get_voices() -> typing.Dict[str, typing.Dict[str, typing.Any]]:
    """Get dict of voices"""
    voices = {}

    # Search for downloaded voices/vocoders
    for voices_dir in voices_dirs:
        if not voices_dir.is_dir():
            continue

        # <LANGUAGE>/<VOICE>-<TTS_SYSTEM>
        for lang_dir in voices_dir.iterdir():
            if (not lang_dir.is_dir()) or (lang_dir.name in VOCODER_DIR_NAMES):
                continue

            # Voice
            voice_lang = lang_dir.name
            for voice_model_dir in lang_dir.iterdir():
                if not valid_voice_dir(voice_model_dir):
                    continue

                full_voice_name = voice_model_dir.name
                voice_name, tts_system = full_voice_name.split("-", maxsplit=1)
                voice_id = f"{voice_lang}/{full_voice_name}"

                voices[voice_id] = {
                    "id": voice_id,
                    "name": voice_name,
                    "language": voice_lang,
                    "tts_system": tts_system,
                    "downloaded": True,
                }

    # Add voices that haven't been downloaded
    for download_name in VOICE_DOWNLOAD_NAMES.values():
        voice_lang, full_voice_name = download_name.split("_", maxsplit=1)
        voice_name, tts_system = full_voice_name.split("-", maxsplit=1)
        voice_id = f"{voice_lang}/{full_voice_name}"

        if voice_id in voices:
            # Already downloaded
            continue

        sample_sentence = _SAMPLE_SENTENCES.get(voice_lang)
        if sample_sentence:
            sample_url = f"https://raw.githubusercontent.com/rhasspy/larynx/master/local/{voice_lang}/{full_voice_name}/samples/{sample_sentence}.wav"
        else:
            # No sample
            sample_url = ""

        voices[voice_id] = {
            "id": voice_id,
            "name": voice_name,
            "language": voice_lang,
            "tts_system": tts_system,
            "downloaded": False,
            "sample_url": sample_url,
        }

    return voices


# -----------------------------------------------------------------------------
# HTTP Endpoints
# -----------------------------------------------------------------------------


@app.route("/api/voices")
async def app_voices() -> Response:
    """Get available voices."""
    return jsonify(get_voices())


@app.route("/api/vocoders")
async def app_vocoders() -> Response:
    """Get available vocoders."""
    vocoders = []

    for voices_dir in voices_dirs:
        if not voices_dir.is_dir():
            continue

        # <VOCODER_SYSTEM>/<VOCODER_MODEL>
        for vocoder_dir in voices_dir.iterdir():
            if (not vocoder_dir.is_dir()) or (
                vocoder_dir.name not in VOCODER_DIR_NAMES
            ):
                continue

            vocoder_system = vocoder_dir.name

            for model_dir in vocoder_dir.iterdir():
                if not valid_voice_dir(model_dir):
                    continue

                model_name = model_dir.name
                vocoder_id = f"{vocoder_system}/{model_name}"

                vocoders.append(
                    {
                        "id": vocoder_id,
                        "name": model_name,
                        "vocoder_system": vocoder_system,
                    }
                )

    return jsonify(vocoders)


def convert_bool(bool_str: str) -> bool:
    """Convert HTML input string to boolean"""
    return bool_str.strip().lower() in {"true", "yes", "on", "1", "enable"}


@app.route("/api/tts", methods=["GET", "POST"])
async def app_say() -> Response:
    """Speak text to WAV."""
    voice = request.args.get("voice", "")
    assert voice, "No voice provided"

    # TTS settings
    noise_scale = request.args.get("noiseScale", args.noise_scale)
    if noise_scale is not None:
        noise_scale = float(noise_scale)

    length_scale = request.args.get("lengthScale", args.length_scale)
    if length_scale is not None:
        length_scale = float(length_scale)

    ssml = convert_bool(request.args.get("ssml", ""))

    # Text can come from POST body or GET ?text arg
    if request.method == "POST":
        text = (await request.data).decode()
    else:
        text = request.args.get("text", "")

    assert text, "No text provided"

    vocoder = request.args.get("vocoder", VocoderQuality.HIGH)

    # Vocoder settings
    denoiser_strength = request.args.get("denoiserStrength", args.denoiser_strength)
    if denoiser_strength is not None:
        denoiser_strength = float(denoiser_strength)

    wav_bytes = await text_to_wav(
        text,
        voice,
        vocoder=vocoder,
        denoiser_strength=denoiser_strength,
        noise_scale=noise_scale,
        length_scale=length_scale,
        ssml=ssml,
    )

    return Response(wav_bytes, mimetype="audio/wav")


@app.route("/api/phonemes", methods=["GET"])
async def api_phonemes():
    """Get phonemes for language"""
    language = request.args.get("language", "en-us")

    phonemes: typing.Dict[str, typing.Dict[str, typing.Any]] = {}
    lang_phonemes = gruut_ipa.Phonemes.from_language(language)
    assert lang_phonemes, f"Unsupported language: {lang_phonemes}"

    for phoneme in lang_phonemes:
        # Try to guess WAV file name for phoneme
        # Files from https://www.ipachart.com/
        wav_path: typing.Optional[Path] = None
        if phoneme.vowel:
            height_str = phoneme.vowel.height.value
            placement_str = phoneme.vowel.placement.value
            rounded_str = "rounded" if phoneme.vowel.rounded else "unrounded"
            wav_path = (
                _WAV_DIR / f"{height_str}_{placement_str}_{rounded_str}_vowel.wav"
            )
        elif phoneme.consonant:
            voiced_str = "voiced" if phoneme.consonant.voiced else "voiceless"
            place_str = phoneme.consonant.place.value.replace("-", "")
            type_str = phoneme.consonant.type.value.replace("-", "_")
            wav_path = _WAV_DIR / f"{voiced_str}_{place_str}_{type_str}.wav"
            if not wav_path.is_file():
                # Try without voicing
                wav_path = _WAV_DIR / f"{place_str}_{type_str}.wav"
        elif phoneme.schwa:
            if phoneme.schwa.r_coloured:
                # Close enough to "r" (er in corn[er])
                wav_path = _WAV_DIR / "alveolar_approximant.wav"
            else:
                # ə
                wav_path = _WAV_DIR / "mid-central_vowel.wav"

        phoneme_dict = {"example": phoneme.example}

        if wav_path and wav_path.is_file():
            # Augment with relative URL to WAV file
            phoneme_dict["url"] = f"wav/{wav_path.name}"
        else:
            _LOGGER.debug("No WAV for %s (%s)", phoneme.text, wav_path)

        phonemes[phoneme.text] = phoneme_dict

    # {
    #   "<PHONEME>": {
    #     "example": "...",
    #     "url": "wav/<voiced>_<place>_<type>.wav"
    #   }
    # }
    return jsonify(phonemes)


@app.route("/api/download", methods=["GET"])
async def api_download():
    """Download voice"""
    voice_id = request.args.get("id", "")
    if "/" in voice_id:
        voice_name = voice_id.split("/", maxsplit=1)[1]
    else:
        voice_name = voice_id

    download_name = VOICE_DOWNLOAD_NAMES.get(voice_name)
    assert download_name, f"No download known for {voice_name}"

    url = args.url_format.format(voice=download_name)
    tts_model_dir = download_voice(voice_name, voices_dirs[0], url)

    return jsonify({"id": voice_id, "url": url, "dir": str(tts_model_dir)})


# -----------------------------------------------------------------------------

# MaryTTS compatibility layer
@app.route("/process", methods=["GET", "POST"])
async def api_process():
    """MaryTTS-compatible /process endpoint"""
    if request.method == "POST":
        data = parse_qs((await request.data).decode())
        text = data.get("INPUT_TEXT", [""])[0]

        if "VOICE" in data:
            voice = data.get("VOICE", [""])[0]
        else:
            voice = data.get("LOCALE", [""])[0]
    else:
        text = request.args.get("INPUT_TEXT", "")
        voice = request.args.get("VOICE", request.args.get("LOCALE", "en-us"))

    # <VOICE>;<VOCODER>
    vocoder: typing.Optional[str] = None

    if ";" in voice:
        voice, vocoder = voice.split(";", maxsplit=1)

    if vocoder is not None:
        # Try to interpret as quality
        vocoder = vocoder.strip()
        vocoder = VOCODER_QUALITY.get(vocoder, vocoder)
    else:
        vocoder = VocoderQuality.HIGH

    # Assume SSML if text begins with an angle bracket
    ssml = text.strip().startswith("<")

    wav_bytes = await text_to_wav(
        text,
        voice,
        vocoder=vocoder,
        denoiser_strength=args.denoiser_strength,
        noise_scale=args.noise_scale,
        length_scale=args.length_scale,
        ssml=ssml,
    )

    return Response(wav_bytes, mimetype="audio/wav")


@app.route("/voices", methods=["GET"])
async def api_voices():
    """MaryTTS-compatible /voices endpoint"""

    load_voices_genders()

    # [voice] [language] [gender] [tech=hmm]
    lines = []

    # Search for downloaded voices/vocoders
    for voices_dir in voices_dirs:
        if not voices_dir.is_dir():
            continue

        # <LANGUAGE>/<VOICE>-<TTS_SYSTEM>
        for lang_dir in voices_dir.iterdir():
            if (not lang_dir.is_dir()) or (lang_dir.name in VOCODER_DIR_NAMES):
                continue

            # Voice
            voice_lang = lang_dir.name
            for voice_model_dir in lang_dir.iterdir():
                if not valid_voice_dir(voice_model_dir):
                    continue

                voice_name_tts = voice_model_dir.name
                full_voice_name = f"{voice_lang}_{voice_name_tts}"
                voice_name, tts_system = voice_name_tts.split("-", maxsplit=1)

                voice_gender = "NA"  # not available
                gender_path = voice_model_dir / "GENDER"

                if gender_path.is_file():
                    # Use GENDER file in voice directory
                    voice_gender = gender_path.read_text().strip()
                else:
                    # Use value from VOICE_GENDERS file
                    voice_gender = VOICE_GENDERS.get(full_voice_name, voice_gender)

                # Repeat voice for each quality level
                for quality in VOCODER_QUALITY:
                    lines.append(
                        f"{voice_name};{quality} {voice_lang} {voice_gender} {tts_system}"
                    )

    return "\n".join(sorted(lines))


@app.route("/version", methods=["GET"])
async def api_version():
    """MaryTTS-compatible /version endpoint"""
    return __version__


# -----------------------------------------------------------------------------


# -----------------------------------------------------------------------------

_CSS_DIR = _DIR / "css"
_IMG_DIR = _DIR / "img"


@app.route("/")
async def app_index():
    """Main page."""
    return await render_template("index.html")


@app.route("/css/<path:filename>", methods=["GET"])
async def css(filename) -> Response:
    """CSS static endpoint."""
    return await send_from_directory(_CSS_DIR, filename)


@app.route("/img/<path:filename>", methods=["GET"])
async def img(filename) -> Response:
    """Image static endpoint."""
    return await send_from_directory(_IMG_DIR, filename)


@app.route("/wav/<path:filename>", methods=["GET"])
async def wav(filename) -> Response:
    """WAV audio static endpoint."""
    return await send_from_directory(_WAV_DIR, filename)


# Swagger UI
api_doc(
    app, config_path=str(_DIR / "swagger.yaml"), url_prefix="/openapi", title="Larynx"
)


@app.errorhandler(Exception)
async def handle_error(err) -> typing.Tuple[str, int]:
    """Return error as text."""
    _LOGGER.exception(err)
    return (f"{err.__class__.__name__}: {err}", 500)


# -----------------------------------------------------------------------------
# Run Web Server
# -----------------------------------------------------------------------------

hyp_config = hypercorn.config.Config()
hyp_config.bind = [f"{args.host}:{args.port}"]

# Create shutdown event for Hypercorn
shutdown_event = asyncio.Event()


def _signal_handler(*_: typing.Any) -> None:
    """Signal shutdown to Hypercorn"""
    shutdown_event.set()


_LOOP.add_signal_handler(signal.SIGTERM, _signal_handler)

try:
    if args.pidfile:
        # Create directory to pidfile
        Path(args.pidfile).parent.mkdir(parents=True, exist_ok=True)
        ctx_pid = pidfile.PIDFile(args.pidfile)
    else:
        # No pidfile
        ctx_pid = contextlib.nullcontext()

    with ctx_pid:
        if args.pidfile:
            _LOGGER.debug("pidfile: %s", args.pidfile)

        # Need to type cast to satisfy mypy
        shutdown_trigger = typing.cast(
            typing.Callable[..., typing.Awaitable[None]], shutdown_event.wait
        )

        _LOOP.run_until_complete(
            hypercorn.asyncio.serve(app, hyp_config, shutdown_trigger=shutdown_trigger)
        )
except KeyboardInterrupt:
    _LOOP.call_soon(shutdown_event.set)
except pidfile.AlreadyRunningError:
    _LOGGER.info("Daemon already running (pidfile=%s). Exiting now.", args.pidfile)
